api/explorer/api/ingest.py

```diff
@@ -98,75 +100,41 @@ def _edge_from_build_input_type(build_input_type: model.BuildInputType) -> graph
         )
 
 
-# TODO: Add retry logic
-def _do_next_graph_level(
-    model_pkg: model.Package,
-    do_fn: Callable[[graph.Package, graph.Package, graph.Edge], None],
-) -> list[model.Package]:
-    # Map package from core model in that used in the graph. This will expand the
-    # input node into N nodes based on the number of the node's output paths.
-    pkgs = safe_parse_package(model_pkg)
-    # Create all dependency nodes and draw an edge between the input package nodes and
-    # them.
-    for pkg in pkgs:
-        for build_input in model_pkg.build_inputs:
-            edge = _edge_from_build_input_type(build_input.build_input_type)
-            try:
-                # Since build_inputs is defined using the core model, we need to parse
-                # it into the graph model similar to what we do above for model_pkg.
-                build_input_pkgs = safe_parse_package(build_input.package)
-            except IngestionError as e:
-                logger.exception(e)
-                build_input_pkgs = []
-            for bi_pkg in build_input_pkgs:
-                do_fn(pkg, bi_pkg, edge)
-    return [bi.package for bi in model_pkg.build_inputs]
-
-
-A = TypeVar("A")
-
-
-def flatten(lls: Iterable[Iterable[A]]) -> list[A]:
-    """Flattens nested iterables into a list"""
-    return [x for sublist in lls for x in sublist]
-
-
-def traverse(
-    nix_graph: model.NixGraph,
-    root_fn: Callable[[graph.Package], None],
-    do_fn: Callable[[graph.Package, graph.Package, graph.Edge], None],
-):
-    """Traverse a Nix graph
-
-    Performs a breadth-first traversal of a nix graph, executing the provided
-    functions on each level of the graph.
+def split_nix_graph(nix_graph: model.NixGraph):
```

Comment:
What does this function do? Maybe a type hint for the type of the return value could help, as well as a docstring.

Reply:
Agreed with Guillaume. Please have a look at this and also feel free to refactor the implementation from my original one if you see ways of improving it to make it easier to read, etc.